src/passes/OptimizeInstructions.cpp

```diff
@@ -1819,6 +1819,24 @@ struct OptimizeInstructions
         }
       }
     }
+    if (curr->type == Type::i32 && Bits::getMaxBits(curr->condition) <= 1 &&
+        Bits::getMaxBits(curr->ifTrue) <= 1 &&
+        Bits::getMaxBits(curr->ifFalse) <= 1) {
+      // The condition and both arms are i32 booleans, which allows us to do
+      // boolean optimizations.
+      Expression* x;
+      Expression* y;
+
+      // x ? y : 0   ==>   x & y
+      if (matches(curr, select(any(&y), ival(0), any(&x)))) {
+        return builder.makeBinary(AndInt32, y, x);
```

Comment:
To avoid moving the operands past each other:
```suggestion
        return builder.makeBinary(AndInt32, x, y);
```
And similar below.

Reply:
I don't think that's right. The order of operands in `select(a, b, c)` is literally `a, b, c`, so we need to keep them that way.

What is confusing here is the loose notation of `x ? y : 0`, which has the condition on the left - but a select (and the `select()` matcher) has it on the right.

Reply:
Oh yes, I was confused by the order changing in the first test.

Reply:
Hmm, but in wasmjs tests x and y looks swapped: https://github.com/WebAssembly/binaryen/pull/4147/commits/9666564dd50dab28c06d119223bc1f26b0ca7b23#diff-dd359c17b1882b7fb4692dfe7e2f371cfbed4a5f80f0c99d868c9a921e451954R25

Perhaps it's canonicalization?

Reply:
Yes, it's the canonicalization mentioned below.

If there was a bug here, the extra test with effects should have caught it. Also I fuzzed over the weekend.

Reply:
Btw `x` and `y` doesn't have side effects due to `select` already should guarantee this. In this case we could swap to more expected `builder.makeBinary(AndInt32, x, y)`. At least for wasm2js

Reply:
It's true that binaryen would not turn an if into a select if the arms have side effects, but someone can create a select with effects:
```wat
(select
  (call $a)
  (call $b)
  (local.get $x))
```
That might not be likely to be emitted by a compiler, but it is possible. This is more realistic:
```wat
(select
  (local.tee $temp
    (call $a))
  (i32.add
    (local.get $temp)
    (i32.const 1))
  (local.get $x))
```